src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,338 @@
+                   lastParameter.Type is IArrayTypeSymbol arrayTypeSymbol &&
+                   arrayTypeSymbol.ElementType.Equals(cancellationTokenType))
+            {
+                IArgumentOperation? paramsArgument = arguments.FirstOrDefault(a => a.ArgumentKind == ArgumentKind.ParamArray);
+                if (paramsArgument?.Value is IArrayCreationOperation arrayOperation)
+                {
+                    // Do not offer a diagnostic if the user already passed a ct to the params
+                    return arrayOperation.Initializer.ElementValues.IsEmpty;
+                }
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(
+            IMethodSymbol method,
+            ITypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out IMethodSymbol? overload)
+        {
+            ImmutableArray<ISymbol> overloads = method.ContainingType.GetMembers(method.Name);
+            System.Collections.Generic.List<IMethodSymbol> ofType = overloads.OfType<IMethodSymbol>().ToList();
+            overload = ofType.FirstOrDefault(methodToCompare =>
+                HasSameParametersPlusCancellationToken(cancellationTokenType, method, methodToCompare));
+
+            return overload != null;
+
+            // Checks if the parameters of the two passed methods only differ in a ct.
+            static bool HasSameParametersPlusCancellationToken(ITypeSymbol cancellationTokenType, IMethodSymbol originalMethod, IMethodSymbol methodToCompare)
+            {
+                // Avoid comparing to itself, or when there are no parameters, or when the last parameter is not a ct
+                if (originalMethod.Equals(methodToCompare.Name) ||
+                    methodToCompare.Parameters.Count(p => p.Type.Equals(cancellationTokenType)) != 1 ||
+                    !methodToCompare.Parameters[^1].Type.Equals(cancellationTokenType))
+                {
+                    return false;
+                }
+
+                IMethodSymbol originalMethodWithAllParameters = originalMethod.ReducedFrom ?? originalMethod.ConstructedFrom ?? originalMethod;
+                IMethodSymbol methodToCompareWithAllParameters = methodToCompare.ReducedFrom ?? methodToCompare.ConstructedFrom ?? methodToCompare;
+
+                // Now compare the types of all parameters before the ct
+                for (int i = 0; i < originalMethodWithAllParameters.Parameters.Length; i++)
+                {
+                    IParameterSymbol? originalParameter = originalMethodWithAllParameters.Parameters[i];
+                    IParameterSymbol? comparedParameter = methodToCompareWithAllParameters.Parameters[i];
+                    if (originalParameter == null || comparedParameter == null || !originalParameter.Type.Equals(comparedParameter.Type))
```

Comment:
@333fred Can the parameters for an `IMethodSymbol` be null? Also can the parameter type be null or will default to ErrorType?

Reply:
Probably not. I removed the `?` and the null comparisons. Feel free to correct me if I'm wrong.